sql/expression/function/time.go

```diff
@@ -828,3 +904,111 @@ func weekFuncLogic(t time.Time) (interface{}, error) {
 	_, wk := t.ISOWeek()
 	return wk, nil
 }
+
+// TimeDiff subtracts the second argument from the first expressed as a time value.
+type TimeDiff struct {
+	expression.BinaryExpression
+}
+
+// NewTimeDiff creates a new NewTimeDiff expression.
+func NewTimeDiff(e1, e2 sql.Expression) sql.Expression {
+	return &TimeDiff{
+		expression.BinaryExpression{
+			Left:  e1,
+			Right: e2,
+		},
+	}
+}
+
+// Type implements the Expression interface.
+func (td *TimeDiff) Type() sql.Type { return sql.Time }
+
+// IsNullable implements the Expression interface.
+func (td *TimeDiff) IsNullable() bool { return false }
+
+func (td *TimeDiff) String() string {
+	return fmt.Sprintf("timediff(%s, %s)", td.Left, td.Right)
+}
+
+// WithChildren implements the Expression interface.
+func (td *TimeDiff) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 2 {
+		return nil, sql.ErrInvalidChildrenNumber.New(td, len(children), 2)
+	}
+	return NewTimeDiff(children[0], children[1]), nil
+}
+
+// Eval implements the Expression interface.
+func (td *TimeDiff) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	// check for valid types
+	if (td.Left.Type() == sql.Time || td.Left.Type() == sql.Timestamp ||
+		td.Left.Type() == sql.Datetime) && (td.Right.Type() == sql.Time ||
+		td.Right.Type() == sql.Timestamp || td.Right.Type() == sql.Datetime) {
+
+		// handle type mismatch
+		if td.Left.Type() != td.Right.Type() {
+			return nil, ErrArgumentTypeMisMatch.New("TIMEDIFF", td.Left, td.Right)
+		}
+
+		left, err := td.Left.Eval(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+
+		if left == nil {
+			return nil, nil
+		}
+
+		leftTime, err := sql.Time.Convert(left)
+		if err != nil {
+			return nil, err
+		}
+
+		right, err := td.Right.Eval(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+
+		if right == nil {
+			return nil, nil
+		}
+
+		rightTime, err := sql.Time.Convert(right)
+		if err != nil {
+			return nil, err
+		}
+
+		if leftTime == nil || rightTime == nil {
+			return nil, nil
+		}
+
+		return toTimeDiff(leftTime.(time.Time), rightTime.(time.Time)), nil
+	}
+	return nil, ErrInvalidArgumentType.New("TIMEDIFF")
+}
+
+func toTimeDiff(left, right time.Time) interface{} {
+	inverted, h, m, s := elapsed(left, right)
+	if inverted {
+		return fmt.Sprintf("%02d:%02d:%09.6f", h, m, s)
+	}
+	return fmt.Sprintf("-%02d:%02d:%09.6f", h, m, s)
+}
```

Comment:
These strings are formatted so that the result is basically an count of hours mins secs diff between times... does this need to be returned as a specific type? besides the Sprintf string type

Reply:
It would probably be easiest for `elapsed` to return a `time.Duration`, as you can then use `sql.Time.Convert` to get the proper value out of it. You can remove `toTimeDiff` altogether, as `sql.Time` will take care of all of that (including returning it as the correct type).